### PR TITLE
Add GeneralSersic2D model

### PR DESCRIPTION
**Pull request by larrybradley**
### Description

This PR adds a `GeneralSersic2D` model, which is a generalized two dimensional Sersic surface brightness profile that allows for "boxy" or "disky" (kite-like) isophote shapes.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/modeling/functional_models.py [hunk 1]
```diff
@@ -3361,6 +3361,207 @@ def _parameter_units_for_data_units(self, inputs_unit, outputs_unit):
...
+        return amplitude * np.exp(-bn * (z ** (1 / n) - 1.0))
+
+    @property
+    def input_units(self):
+        if self.x_0.input_unit is None:
+            return None
+        return {
+            self.inputs[0]: self.x_0.input_unit,
+            self.inputs[1]: self.y_0.input_unit,
+        }
```
Comment (WilliamJamieson) [inline astropy/modeling/functional_models.py:3548]:
```suggestion
        if self.x_0.input_unit is not None:
            return {
                self.inputs[0]: self.x_0.input_unit,
                self.inputs[1]: self.y_0.input_unit,
            }
```

File: astropy/modeling/functional_models.py [hunk 2]
```diff
@@ -3361,6 +3361,207 @@ def _parameter_units_for_data_units(self, inputs_unit, outputs_unit):
...
+
+    @classmethod
+    def evaluate(cls, x, y, amplitude, r_eff, n, x_0, y_0, c, ellip, theta):
+        """Two dimensional Sersic profile function."""
+        if cls._gammaincinv is None:
+            from scipy.special import gammaincinv
+
+            cls._gammaincinv = gammaincinv
+
+        bn = cls._gammaincinv(2.0 * n, 0.5)
```
Comment (WilliamJamieson) [inline astropy/modeling/functional_models.py:3530]:
```suggestion
    @staticmethod
    def evaluate(cls, x, y, amplitude, r_eff, n, x_0, y_0, c, ellip, theta):
        """Two dimensional Sersic profile function."""
        from scipy.special import gammaincinv

        bn = gammaincinv(2.0 * n, 0.5)
```

File: astropy/modeling/functional_models.py [hunk 3]
```diff
@@ -3361,6 +3361,207 @@ def _parameter_units_for_data_units(self, inputs_unit, outputs_unit):
...
+        a, b = r_eff, (1 - ellip) * r_eff
+        cos_theta, sin_theta = np.cos(theta), np.sin(theta)
+        x_maj = np.abs((x - x_0) * cos_theta + (y - y_0) * sin_theta)
+        x_min = np.abs(-(x - x_0) * sin_theta + (y - y_0) * cos_theta)
+
+        expon = 2.0 + c
+        inv_expon = 1.0 / expon
+        z = ((x_maj / a) ** expon + (x_min / b) ** expon) ** inv_expon
```
Comment (WilliamJamieson) [inline astropy/modeling/functional_models.py:3538]:
```suggestion
        cos_theta = np.cos(theta)
        sin_theta = np.sin(theta)
        x_maj = np.abs((x - x_0) * cos_theta + (y - y_0) * sin_theta)
        x_min = np.abs(-(x - x_0) * sin_theta + (y - y_0) * cos_theta)

        b = (1 - ellip) * r_eff
        expon = 2.0 + c
        inv_expon = 1.0 / expon
        z = ((x_maj / r_eff) ** expon + (x_min / b) ** expon) ** inv_expon
```

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]
This might need a "what's new" blurb.

**Comment by WilliamJamieson** [comment]
> * `GeneralSersic2D` and `Sersic2D` formulae in the docs look almost identical.

I agree, one should be able to just add the parameter `c` with default value = 0.

File: astropy/modeling/functional_models.py [hunk 4]
```diff
@@ -3361,6 +3361,207 @@ def _parameter_units_for_data_units(self, inputs_unit, outputs_unit):
...
+        # defined.
+        if inputs_unit[self.inputs[0]] != inputs_unit[self.inputs[1]]:
+            raise UnitsError("Units of 'x' and 'y' inputs should match")
+        return {
+            "x_0": inputs_unit[self.inputs[0]],
+            "y_0": inputs_unit[self.inputs[0]],
+            "r_eff": inputs_unit[self.inputs[0]],
+            "theta": u.rad,
+            "amplitude": outputs_unit[self.outputs[0]],
+        }
```
Comment (WilliamJamieson) [inline astropy/modeling/functional_models.py:3562]: I believe that `c` also needs to have a unit returned.

**Comment by larrybradley** [comment]
I didn't add the `c` parameter to `Sersic2D` because it would break existing code.  I'm also not keen on having `c` be fixed because no other model does that, and the principle of least surprise says it's going to lead to confusion (even if explained in the docs).  Also there are other existing models that could be thought of as special cases of other models (e.g., some in the power-law models).

Let me know if you still prefer adding `c` to `Sersic2D` and I'll make that change.

File: astropy/modeling/functional_models.py (hunk 4 above)
Comment (larrybradley) [reply astropy/modeling/functional_models.py:3562]: No, `c` is a unitless exponent.

File: astropy/modeling/functional_models.py (hunk 2 above)
Comment (larrybradley) [reply astropy/modeling/functional_models.py]: I also didn't understand why `_gammaincinv` was defined.  Fixed.

File: astropy/modeling/functional_models.py (hunk 1 above)
Comment (larrybradley) [reply astropy/modeling/functional_models.py]: indeed.  again that's code copied from `Sersic2D`. fixed.

File: astropy/modeling/functional_models.py (hunk 3 above)
Comment (larrybradley) [reply astropy/modeling/functional_models.py]: fixed copied code here too.

File: astropy/modeling/functional_models.py [hunk 5]
```diff
@@ -3361,6 +3359,166 @@ def _parameter_units_for_data_units(self, inputs_unit, outputs_unit):
...
+    ellip = Parameter(default=0, description="Ellipticity")
+    theta = Parameter(
+        default=0.0,
+        description=(
+            "Rotation angle either as a float (in radians) or a |Quantity| angle"
+        ),
+    )
+    c = Parameter(default=0, description="Isophote shape parameter")
+
+    _gammaincinv = None
```
Comment (WilliamJamieson) [inline astropy/modeling/functional_models.py:3519]: This snuck in.
```suggestion
```

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

**Comment by WilliamJamieson** [comment]
Only other thing is to update the docs as @pllim suggested: https://github.com/astropy/astropy/pull/15545#pullrequestreview-1702322220

**Comment by larrybradley** [comment]
I've update the docs as suggested by @pllim and added a "What's new" entry.

File: astropy/modeling/functional_models.py (hunk 5 above)
Comment (larrybradley) [reply astropy/modeling/functional_models.py]: fixed, thanks.

**Review by WilliamJamieson — APPROVED** [review]

**Comment by pllim** [comment]
Thanks!

**Pull request merged**